src/distdl/nn/batch_norm.py

```diff
@@ -0,0 +1,227 @@
+import torch
+
+from distdl.backends.mpi.tensor_comm import assemble_global_tensor_structure
+from distdl.nn.broadcast import Broadcast
+from distdl.nn.module import Module
+from distdl.nn.sum_reduce import SumReduce
+from distdl.utilities.slicing import compute_start_index
+from distdl.utilities.slicing import compute_stop_index
+from distdl.utilities.slicing import worker_layout
+from distdl.utilities.torch import zero_volume_tensor
+
+
+class DistributedBatchNorm(Module):
+    r"""A distributed batch norm layer.
+
+    Applies Batch Normalization using mini-batch statistics.
+    This layer is a distributed and generalized version of the PyTorch BatchNormNd layers.
+    Currently, parallelism is supported in all dimensions except the feature dimension (dimension 2).
+
+    Parameters
+    ----------
+    P_x :
+        Partition of the input tensor.  Outputs are of the same shape,
+        and therefore re-use the input partition.
+    num_features :
+        Number of features in the input.
+        For example, this should equal C in an input of shape (N, C, L).
+    eps : optional
+        A value added to the denominator for numerical stability.
+        Default is 1e-5.
+    momentum : optional
+        The value used for the running_mean and running_var computation.
+        Can be set to None for cumulative moving average (i.e. simple average).
+        Default is 0.1.
+    affine : optional
+        a boolean value that when set to True, this module has learnable affine parameters.
+        Default is True.
+    track_running_stats : optional
+        a boolean value that when set to True, this module tracks the running mean and variance,
+        and when set to False, this module does not track such statistics and uses batch statistics
+        instead in both training and eval modes if the running mean and variance are None.
+        Default is True.
+    """
+
+    def __init__(self, P_x,
+                 num_features, eps=1e-05, momentum=0.1, affine=True,
+                 track_running_stats=True):
+        super(DistributedBatchNorm, self).__init__()
+        self.num_dimensions = len(P_x.shape)
+        assert self.num_dimensions >= 2
+        self.num_features = num_features
+        self.eps = eps
+        self.momentum = momentum
+        self.affine = affine
+        self.track_running_stats = track_running_stats
+        self.inputs_seen = 0
+
+        # Determine the size of the local trainable parameters (this is a bit of a hack)
```

Comment:
Would it be helpful if I moved this into a helper function that just determines the data split sizes across all dimensions (given the partition and a shape)? I think there's some chance this scenario will arise in the future, and it's a fairly fast bookkeeping computation.

Reply:
That would not be outside of the realm of reason.  It would require more than the partition and shape though, as the result can be unbalanced.

The main simplification is that it can be computed on channels only, no need to assemble the whole thing then select just the channel result.